# low_bismap_methyl_calls: Correcting Methylation Calls in Low-Mappability Regions
Tools to assess called methylation in low mappability regions (as defined by bismap). Aligner MapQ does not always correspond with low mappability.

![ideogram](./chroms_ideogram.png)



For the Nextflow pipeline used to examine the effectiveness of this method, see the workflow in miscalls_full.nf, and the workflows it imports from miscalls_pair.nf and miscalls_one.nf

## Running the Nextflows

### Files needed

* ClinVar VCF
* GENCODE GFF3
* hg38/GRCh38 Bismap multi-read k100 bigWig. This is not the bedGraph on the Bismap site, get it from the Genome Browser track.
* T2T Bismap multi-read k100 bigWig. This is not available and must be generated by running umap.
* uBAM or FASTQ files for all sequencing reads (aligned BAMs will also work, but will be treated as uBAMs). They do not need to be adapter-trimmed in advance, the realign pipeline does so using fastp.
* FASTA reference sequences for GRCh38 and T2T, with FAI indices

### Realignment

Run realign.nf as follows:

`nextflow run realign.nf --refs "/path/to/grch38/ref.fa,/path/to/t2t/ref.fa" --input_path "/path/to/dir/with/seqs" --suffixes "grch38,t2t" --temp_dir "/path/to/tmp/dir" --output_path "/path/to/output/dir" -w "/path/to/work/dir" --data_path "/path/to/dir/to/store/aligner/indices" -c "nextflow.config"`

The order of the `--refs` paths needs to match the order of the suffixes passed to `--suffixes`, or the files will be misnamed, but there is no requirement they be in the order grch38,t2t as here.

The sequence data in the directory passed to `--input_path` (hereafter called `input_dir`)` *must* be arranged as follows:

1. Each dataset/library must be a subdirectory under `input_dir`.
2. Within each subdirectory, there must either be two FASTQs, named `[something]_1.fastq` and `[something]_2.fastq`, or a BAM named `[something].bam` and an index named `[something].bam.bai`.

### Analysis prep

For the analysis, the GENCODE and ClinVar files must be further processed to obtain the "ClinVar regions BEDs". This is done using clinvar.nf. This pipeline must be run twice, once for GRCh38 and once for T2T, as follows:

`nextflow run clinvar.nf --clinvar "/path/to/clinvar.vcf" --gencode "/path/to/gencode.gff3" --output_path "/path/to/save/output" -c "nextflow.config"`

It is recommended to specify a different `--output_path` option for the runs so they save the ClinVar regions BEDs to different directories for each reference.

For performance, the MethylDackel invocation in this pipeline uses the custom .bbm file format, so preprocessing is required to convert the bigWig files to .bbm as follows for each bigWig file:

`MethylDackel extract -M <mappability bigWig> -O`

### Analysis

[TODO]


